source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -434,6 +445,13 @@ ReadOrParseState Filter::readProxyHeader(Network::IoHandle& io_handle) {
     if (nread < 1) {
       ENVOY_LOG(debug, "failed to read proxy protocol (no bytes read)");
       return ReadOrParseState::Error;
+    } else if (nread < PROXY_PROTO_V2_HEADER_LEN &&
+               config_.get()->allowRequestsWithoutProxyProtocol()) {
+      if (nread < PROXY_PROTO_V1_SIGNATURE_LEN ||
```

Comment:
Although it is rare case, there still have the chance of receiving the partial v1 signature, like just a 'P' char, the left of the signature received by the second `recv` call.  But this check will just skip the left of signature checking.

we could return `SkipFilterError` at https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abL446
Then it will got all the data, and compared the signature of v1 and v2.

But if we only receive very short data from the client (short than both v1 and v2 signature), then I feel we only can waiting for a timeout of listener filter. But that is bad for the usecase of issue.

Another idea is if the `nread` is less than v1 and v2 signature, then we compare part of signature, like if the first byte is `X', then compare to the first byte of v1 signature, it is `P`. Also compare the first byte of v2 signature, then we know it can't be the proxy protocol.

Reply:
@soulxu can you update the link from

> we could return SkipFilterError at https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abL446
Then it will got all the data, and compared the signature of v1 and v2.

to a link to main / a commit and where you're proposing this change? I think with the recent commits to this PR that the link provided is dated, I'm not sure where you're proposing we return the error

Reply:
> Another idea is if the nread is less than v1 and v2 signature, then we compare part of signature, like if the first byte is X', then compare to the first byte of v1 signature, it is P`. Also compare the first byte of v2 signature, then we know it can't be the proxy protocol.

happy to implement this idea, makes sense to me. would this be sufficient or did you still want to discuss your earlier proposal (with the link to code that I haven't been able to follow yet)

Reply:
also related https://github.com/envoyproxy/envoy/pull/18951#discussion_r746888896